### PR TITLE
transcomb new recipe

### PR DESCRIPTION
**Pull request by mictadlo**
* [x] I have read the guidelines above.
* [x] This PR adds a new recipe.
* [ ] This PR updates an existing recipe.
* [ ] This PR does something else (explain below).

### CODE REVIEW

**Comment by mictadlo** [comment]
@bioconda/all How can I replace text in the same way for Linux and Mac, because I ran [into](https://travis-ci.org/bioconda/bioconda-recipes/jobs/191515353#L612)?

**Comment by jerowe** [comment]
`sed -i.bak 's/thing/otherthing/' MyFile`

I think that does it. I had the same problem before. ;-)

**Comment by mictadlo** [comment]
@jerowe:  I still have a [problem](https://travis-ci.org/bioconda/bioconda-recipes/jobs/191534370#L612)

**Comment by nsoranzo** [comment]
> sed: -e expression #1, char 40: unterminated `s' command

You need a / at the end of the sed commands, e.g.

sed -i.bak -e 's/foo/bar/' filename

File: recipes/transcomb/build.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+mkdir -p $PREFIX/bin
+
+
+sed -i.bak 's/<Path_To_BOOST_LIB_DIR>/${PREFIX}\/lib' src/CMakeList.txt
+sed -i.bak 's/<Path_To_BOOST_INCLUDE_DIR>/${PREFIX}\/include' src/CMakeList.txt
+sed -i.bak 's/<Path_To_BAMTOOLS_INCLUDE_DIR>/${PREFIX}\/include' src/CMakeList.txt
+sed -i.bak 's/<Path_To_BAMTOOLS_LIB_DIR>/${PREFIX}\/lib' src/CMakeList.txt
```
Comment (sebastian-luna-valero) [inline recipes/transcomb/build.sh]: @mictadlo try:
```
sed -i'' -e '/<Path_To_BOOST_LIB_DIR>/${PREFIX}\/lib' src/CMakeList.txt
sed -i'' -e '/<Path_To_BOOST_INCLUDE_DIR>/${PREFIX}\/include' src/CMakeList.txt
sed -i'' -e '/<Path_To_BAMTOOLS_INCLUDE_DIR>/${PREFIX}\/include' src/CMakeList.txt
sed -i'' -e '/<Path_To_BAMTOOLS_LIB_DIR>/${PREFIX}\/lib' src/CMakeList.txt
```

**Comment by mictadlo** [comment]
Thank you.`Sed` worked, but it did not replace `${PREFIX}` and therefore I think it can't find boost.

**Comment by ksuderman** [comment]
Take care when using sed to patch paths on Linux/OSX systems as any '/' characters in the path will be mistaken by sed to be the '/' that terminates the replacement pattern; (e.g. the '/' in ${PREFIX}/lib). Fortunately, sed allows for any character to be used to demarcate the pattern/replacement.  The following should work:

```bash
sed -i .bak 's|<Path_To_BOOST_LIB_DIR>|${PREFIX}/lib|g' src/CMakeList.txt
```

Note the use of '|' (pipe) instead of '/' in the sed command.

**Comment by mictadlo** [comment]
This makes it easier, but I still get `set(BOOST_LIB_DIR ${PREFIX}/lib)` with @ksuderman solution. How could `${PREFIX}/lib` show the actual `PATH` to `lib`?

**Comment by mictadlo** [comment]
It fail to build, because of 
```
[ 71%] Building CXX object CMakeFiles/Pre_Alignment.dir/Pre_Alignment.cc.o
/anaconda/conda-bld/transcomb_1484536138920/work/TransComb_v.1.0/src/Pre_Alignment.cc:4:10: fatal error: 
      'boost/random.hpp' file not found
#include <boost/random.hpp>
```

**Comment by jerowe** [comment]
@mictadlo, it wasn't building because the CMakeLists.txt is not paying attention to the usual boost environmental variables. Boost is kind of a pain to compile against, but for the most part I have been able to get it with this.

This should be your build.sh

```
#!/bin/bash

set -x -e

mkdir -p "${PREFIX}/bin"

export INCLUDE_PATH="${PREFIX}/include/:${PREFIX}/include/bamtools/"
export LIBRARY_PATH="${PREFIX}/lib"
export LD_LIBRARY_PATH="${PREFIX}/lib"
export BOOST_INCLUDE_DIR="${PREFIX}/include"
export BOOST_LIBRARY_DIR="${PREFIX}/lib"

#export LIBS='-lboost_system -lboost_program_options -lboost_filesystem -lboost_timer'

export BAMTOOLS_INCLUDE_DIR="${PREFIX}/include/bamtools/"
export BAMTOOLS_LIBRARY_DIR="${PREFIX}/lib/"

export CXXFLAGS="-DUSE_BOOST -I${BOOST_INCLUDE_DIR} -L${BOOST_LIBRARY_DIR}"
export LDFLAGS="-L${BOOST_LIBRARY_DIR} -lboost_filesystem -lboost_system"

mkdir build
cd build
cmake ../src
make

cd ..

cp TransComb $PREFIX/bin
cp build/Assemble $PREFIX/bin
cp build/CorrectName $PREFIX/bin
cp build/Pre_Alignment $PREFIX/bin
```

Also, CMake deals with its environmental variables differently than just a Makefile. The syntax is $ENV{My_Env_VAR}.

Additionally, we ask that everyone pins boost in their meta.yml.

meta.yaml
```
{% set name = "TransComb" %}
{% set version = "1.0" %}

package:
  name: {{ name|lower }}
  version: {{ version }}

build:
  number: 0
  string: boost{{CONDA_BOOST}}_{{PKG_BUILDNUM}}

source:
  fn: {{ name|lower }}_{{ version }}.tar.gz  
  url: http://downloads.sourceforge.net/project/transcriptomeassembly/TransComb_v.1.0.tar.gz
  md5: 2bae37e0e59646f900dd0683f572af9b
  patches:
    - CMakeLists.txt.patch

requirements:
  build:
  - gcc # [not osx]
  - llvm  # [osx]
  - boost {{CONDA_BOOST}}*
  - bamtools
  - samtools ==0.1.19
  - zlib
  - cmake

  run:
  - libgcc # [not osx]
  - zlib
  - boost {{CONDA_BOOST}}*
  - bamtools
  - samtools

test:
  commands:
    - TransComb -h | grep "TransComb -b file.bam -s unstranded"
    - CorrectName -h | grep "CorrectName fq reads_1.fq reads_2.fq"
    
    #Neither one of these has useful output, so skipping tests
    #- Assemble
    #- Pre_Alignment -h

about:
  home: https://github.com/yechengxi/SparseAssembler
  summary: A sparse k-mer graph based, memory-efficient genome assembler
```

**Comment by jerowe** [comment]
Lastly, if you don't have to don't even bother with sed. I find its much easier to just make a patch file.

src/CMakeLists.txt.patch
```
--- src.CMakeLists.txt  2017-01-17 09:36:04.786802923 +0400
+++ src/CMakeLists.txt  2017-01-17 09:38:13.508652701 +0400
@@ -2,20 +2,20 @@
 project (Pre_Alignment)
 
 set(CMAKE_CXX_COMPILER "g++")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3")
+set(CMAKE_CXX_FLAGS "$ENV{CMAKE_CXX_FLAGS} -O3")
 
 #Warning: The followings need your settings...(we have shown some examples below them.)
-set(BOOST_LIB_DIR <Path_To_BOOST_LIB_DIR>)
+set(BOOST_LIB_DIR $ENV{BOOST_LIBRARY_DIR})
 #Example: set(BOOST_LIB_DIR /home/juntaosdu/local/boost/lib)
-set(BOOST_INCLUDE_DIR <Path_To_BOOST_INCLUDE_DIR>)
+set(BOOST_INCLUDE_DIR $ENV{BOOST_INCLUDE_DIR})
 #Example: set(BOOST_INCLUDE_DIR /home/juntaosdu/local/boost/include)
-set(BAMTOOLS_LIB_DIR <Path_To_BAMTOOLS_LIB_DIR>)
+set(BAMTOOLS_LIB_DIR $ENV{BAMTOOLS_LIBRARY_DIR})
 #Example: set(BAMTOOLS_LIB_DIR /home/juntaosdu/bamtools/lib)
-set(BAMTOOLS_INCLUDE_DIR <Path_To_BAMTOOLS_INCLUDE_DIR>)
+set(BAMTOOLS_INCLUDE_DIR $ENV{BAMTOOLS_INCLUDE_DIR})
 #Example: set(BAMTOOLS_INCLUDE_DIR /home/juntaosdu/bamtools/include)
 
-include_directories(. ${BOOST_INCLUDE_DIR} ${BAMTOOLS_INCLUDE_DIR})
-link_directories(${BOOST_LIB_DIR} ${BAMTOOLS_LIB_DIR})
+include_directories(. $ENV{BOOST_INCLUDE_DIR} $ENV{BAMTOOLS_INCLUDE_DIR})
+link_directories($ENV{BOOST_LIB_DIR} $ENV{BAMTOOLS_LIB_DIR})
 
 #add_executable(Assemble Graph_NH_old_combine_filter.cc expression_level.h
 #Find_junc_last_map.h Insert.h Sort.h edge_node_pair.h False_junc.h sep_gene.h
```

**Comment by jerowe** [comment]
I have a branch that build on my local docker [here](https://github.com/jerowe/bioconda-recipes/tree/feature/transcomb/recipes/transcomb).

**Comment by mictadlo** [comment]
I could close this PR and you could submit your solution. Which you did very well.

**Comment by jerowe** [comment]
I like to spread the 'how to compile with boost because its a pain in the butt' knowledge around. ;-)

**Pull request closed without merging**